### PR TITLE
Fix switch expression patterns to have consistent behaviour

### LINKED ISSUE #3070: Fix switch expression patterns to have consistent behaviour
Currently, CFG built by the csharp plugin is not consistent related to switch expression patterns. This makes it hard to predict behaviour of symbolic execution when the number of symbolic values pushed on the stack may differ depending on the pattern type.

### PR DESCRIPTION
**Pull request by christophe-zurn-sonarsource**
Fixes #3070

### CODE REVIEW

File: sonaranalyzer-dotnet/src/SonarAnalyzer.CSharp/SymbolicExecution/CSharp/CSharpExplodedGraph.cs [hunk 1]
```diff
@@ -769,16 +769,17 @@ private void VisitCaseSwitchBinaryBranchBlock(BinaryBranchBlock branchBlock, Exp
         private void VisitSwitchExpressionArmBinaryBranch(BinaryBranchBlock branchBlock, ExplodedGraphNode node, SwitchExpressionArmSyntaxWrapper armSyntaxWrapper)
```
Comment (costin-zaharia-sonarsource) [inline sonaranalyzer-dotnet/src/SonarAnalyzer.CSharp/SymbolicExecution/CSharp/CSharpExplodedGraph.cs]: One thing that I find a bit confusing now is that we handle the constant pattern differently than RecursivePattern. Constant pattern is processed already when building the CFG and it will appear as Identifier in the CFG but the RecursivePattern will remain as it is.

It's just a small thing and avoiding the additional changes keeps the PR small which is nice, but I had to mention it 😄

File: sonaranalyzer-dotnet/tests/SonarAnalyzer.UnitTest/ControlFlowGraph/ControlFlowGraphTest.cs [hunk 2]
```diff
@@ -5018,6 +5013,30 @@ public void Cfg_PropertyPatternClause_InsideIf()
...
+            VerifyAllInstructions(entryBlock,
+                "address",
+                "address is Address _"); // IsPatternExpression
+
+            trueBlock.SuccessorBlock.Should().Be(exitBlock);
+            VerifyAllInstructions(trueBlock, "true");
+
+            VerifyNoInstruction(exitBlock);
+        }
+
```
Comment (costin-zaharia-sonarsource) [inline sonaranalyzer-dotnet/tests/SonarAnalyzer.UnitTest/ControlFlowGraph/ControlFlowGraphTest.cs]: Is this related to the current PR?

File: sonaranalyzer-dotnet/tests/SonarAnalyzer.UnitTest/SymbolicExecution/ExplodedGraphTests.cs [hunk 3]
```diff
@@ -722,6 +708,108 @@ public int Main(string str)
             context.WalkWithExitBlocks(7, 3);
         }
 
+        [TestMethod]
+        [TestCategory("Symbolic execution")]
+        public void ExplodedGraph_SwitchExpression_AllPatterns()
+        {
+            const string testInput = @"string str = GetStr(); var x = str switch { null => 1, int i => 2, { } => 3, _ => 4};";
+            var context = new ExplodedGraphContext(testInput);
+            var strParameter = context.MainMethod.ParameterList.Parameters.First();
```
Comment (costin-zaharia-sonarsource) [inline sonaranalyzer-dotnet/tests/SonarAnalyzer.UnitTest/SymbolicExecution/ExplodedGraphTests.cs]: **strParameter** is not used

File: sonaranalyzer-dotnet/tests/SonarAnalyzer.UnitTest/SymbolicExecution/ExplodedGraphTests.cs [hunk 4]
```diff
@@ -722,6 +708,108 @@ public int Main(string str)
             context.WalkWithExitBlocks(7, 3);
         }
 
+        [TestMethod]
+        [TestCategory("Symbolic execution")]
+        public void ExplodedGraph_SwitchExpression_AllPatterns()
```
Comment (costin-zaharia-sonarsource) [inline sonaranalyzer-dotnet/tests/SonarAnalyzer.UnitTest/SymbolicExecution/ExplodedGraphTests.cs]: All constant patterns? 😄

File: sonaranalyzer-dotnet/tests/SonarAnalyzer.UnitTest/SymbolicExecution/ExplodedGraphTests.cs [hunk 5]
```diff
@@ -722,6 +708,108 @@ public int Main(string str)
...
+            assignmentInspected.Should().Be(4);
+        }
+
+        [TestMethod]
+        [TestCategory("Symbolic execution")]
+        public void ExplodedGraph_SwitchExpression_NullConstraint()
+        {
+            const string testInput = @"const string a = null; string str = GetStr(); var x = str switch { a => 1, _ => 2};";
+            var context = new ExplodedGraphContext(testInput);
+            var strParameter = context.MainMethod.ParameterList.Parameters.First();
```
Comment (costin-zaharia-sonarsource) [inline sonaranalyzer-dotnet/tests/SonarAnalyzer.UnitTest/SymbolicExecution/ExplodedGraphTests.cs]: **strParameter** is not used

File: sonaranalyzer-dotnet/src/SonarAnalyzer.CFG/ControlFlowGraph/CSharpControlFlowGraphBuilder.cs [hunk 6]
```diff
@@ -523,31 +523,14 @@ private Block BuildPatternExpression(PatternSyntaxWrapper patternSyntaxWrapper,
             }
             else if (RecursivePatternSyntaxWrapper.IsInstance(patternSyntaxWrapper))
             {
-                return BuildRecursivePatternExpression((RecursivePatternSyntaxWrapper)patternSyntaxWrapper, currentBlock);
+                // The recursive pattern will be handled in CSharpExplodedGraph and UcfgInstructionFactory.
+                currentBlock.ReversedInstructions.Add(patternSyntaxWrapper);
```
[the remark this answers is not in the document]
Comment (christophe-zurn-sonarsource) [reply sonaranalyzer-dotnet/src/SonarAnalyzer.CFG/ControlFlowGraph/CSharpControlFlowGraphBuilder.cs]: Indeed. This is because we have this code:
```
                case SyntaxKindEx.RecursivePattern: // https://github.com/SonarSource/sonar-dotnet/issues/2937
                case SyntaxKindEx.ConstantPattern:
                    // The 0 in 'case 0 when ...'
                    // Do nothing
                    break;
```
in the `CSharpExplodedGraph` visit instructions method.

File: sonaranalyzer-dotnet/tests/SonarAnalyzer.UnitTest/ControlFlowGraph/ControlFlowGraphTest.cs (hunk 2 above)
Comment (christophe-zurn-sonarsource) [reply sonaranalyzer-dotnet/tests/SonarAnalyzer.UnitTest/ControlFlowGraph/ControlFlowGraphTest.cs]: Nope. It remained from when I was testing several cases from the SE

File: sonaranalyzer-dotnet/tests/SonarAnalyzer.UnitTest/SymbolicExecution/ExplodedGraphTests.cs (hunk 4 above)
Comment (christophe-zurn-sonarsource) [reply sonaranalyzer-dotnet/tests/SonarAnalyzer.UnitTest/SymbolicExecution/ExplodedGraphTests.cs]: Nope, this actually tests all kind of patterns:
`int i` is `VariableDeclarationPattern`,
`{ }` is (empty) `RecursivePattern`,
`_` is `DiscardPattern`
